src/engine/sparqlExpressions/SparqlExpressionValueGetters.cpp

```diff
@@ -252,7 +256,37 @@ std::optional<std::string> LanguageTagValueGetter::operator()(
     case Double:
     case Date:
     case Undefined:
-      return std::nullopt;
+      if constexpr (std::is_same_v<T, IdOrLiteralOrIri>) {
+        return Id::makeUndefined();
+      } else {
+        return std::nullopt;
+      }
   }
   AD_FAIL();
 }
+
+//_____________________________________________________________________________
+IdOrLiteralOrIri IriOrUriValueGetter::operator()(
+    ValueId id, const EvaluationContext* context) const {
+  return getValue<IdOrLiteralOrIri>(id, context, *this);
+}
+
+//______________________________________________________________________________
+std::optional<std::string> LanguageTagValueGetter::operator()(
+    ValueId id, const EvaluationContext* context) const {
+  return getValue<std::optional<std::string>>(id, context, *this);
+}
+
+//______________________________________________________________________________
+IdOrLiteralOrIri IriOrUriValueGetter::operator()(
+    const LiteralOrIri& litOrIri,
+    [[maybe_unused]] const EvaluationContext* context) const {
+  const auto& optIri = Iri::fromIrirefWithBasicUrlCheck(
```

Comment:
I think for now you should simply return the content without any normalization.
The whole resolving against base IRI stuff is something that QLever currently doesn't support at all, so we should implement it in one go at a later time.
In particlar QLever currently supports IRIs that are not valid IRIs like `<is-a>` etc.

The only thing that is worth considering (check what happens there), is if a string contains special characters that are not allowed in IRIs. (which are those, should we normalize them), but there I also would have to check on how to properly normalize or escape them (percent encoding? anything else?)

Reply:
Ah i see, i didn't realize that `BASE` was a declaration similar to a `PREFIX` at the beginning of a query.

Given that we implement `IRI()` and `URI()` in combination, and `IRIs` allow a wider range of characters, if we normalize at all, probably only w.r.t. the IRI range.

Reply:
`BASE` is already defined by the parser grammar, but not implemented yet. Resolving against `BASE` could be maybe done with help of `sparqlExpression::EvaluationContext`.